### PR TITLE
PRO-6657: Following values for dynamic choices

### PR DESCRIPTION
**Pull request by boutell**
## Summary

Support for "following:" syntax with dynamic choice functions, which receive the values they follow as a third argument containing an object. Note the property names do include the "<" if following higher levels. See matching branch of testbed for examples.

## What are the specific steps to test this change?

See testbed.

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [X] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [X] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [ ] The changelog is updated
- [ ] Related documentation has been updated
- [ ] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: modules/@apostrophecms/schema/index.js [hunk 1]
```diff
@@ -1698,12 +1706,87 @@ module.exports = {
...
-          self.register('widget', type, manager.schema);
-        });
+        self.schemaPointers = {};
+        registerMetaType(self.apos.doc.managers, 'doc');
+        registerMetaType(self.apos.area.widgetManagers, 'widget');
+        function registerMetaType(managers, metaType) {
+          for (const [ type, manager ] of Object.entries(managers)) {
+            const schema = manager.schema;
+            self.register(metaType, type, schema);
+            const pointer = {
```
Comment (boutell) [inline modules/@apostrophecms/schema/index.js:1716]: All this pointer stuff allows us to resolve from relative field names to field ids and then the actual field object, without trusting the client.

File: modules/@apostrophecms/schema/index.js [hunk 2]
```diff
@@ -1779,36 +1878,66 @@ module.exports = {
...
+        const field = self.getFieldById(fieldId);
+        let choices = [];
+        if (
+          !field ||
+          !self.fieldTypes[field.type].dynamicChoices ||
+          !(field.choices && typeof field.choices === 'string')
+        ) {
+          throw self.apos.error('invalid');
+        }
+        if (field.following) {
```
Comment (boutell) [inline modules/@apostrophecms/schema/index.js:1897]: We have to validate the fields being followed.

File: modules/@apostrophecms/schema/index.js [hunk 3]
```diff
@@ -1779,36 +1878,66 @@ module.exports = {
...
+          for (const follows of field.following) {
+            const relatedField = self.getFieldByRelativePath(field._id, follows);
+            relatedField.if = undefined;
```
Comment (boutell) [inline modules/@apostrophecms/schema/index.js:1900]: We ease up the requirements as the form may be incomplete.

File: modules/@apostrophecms/schema/index.js [hunk 4]
```diff
@@ -1828,6 +1957,11 @@ module.exports = {
             throw self.apos.error('invalid', error.message);
           }
         }
+      },
+      post: {
```
Comment (boutell) [inline modules/@apostrophecms/schema/index.js:1961]: Preferable because there could be a lot of following data.

**Review by myovchev — APPROVED** [review]
LGTM

File: modules/@apostrophecms/schema/lib/addFieldTypes.js [hunk 5]
```diff
@@ -861,7 +861,7 @@ module.exports = (self) => {
 
   self.addFieldType({
     name: 'object',
-    async convert(req, field, data, destination, { fetchRelationships = true, doc = {} } = {}) {
+    async convert(req, field, data, destination, { fetchRelationships = true, ancestors = {}, doc = {} } = {}) {
```
Comment (myovchev) [inline modules/@apostrophecms/schema/lib/addFieldTypes.js:864]: It causes an eslint warning.

**Review by myovchev — APPROVED** [review]

**Pull request merged**